File: packages/reown_yttrium/android/src/main/kotlin/com/reown/reown_yttrium/Ton.kt
package com.reown.reown_yttrium

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log

import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import uniffi.yttrium.Keypair
import uniffi.yttrium.PulseMetadata
import uniffi.yttrium.TonClient
import uniffi.yttrium.TonClientConfig
import kotlin.collections.get

/**
 * Ton.kt
 *
 * Supports multiple TonClient instances (e.g. mainnet, testnet).
 * Use clientId (e.g. "mainnet", "testnet") to identify each client.
 */
class Ton(private val projectId: String, private val networkId: String, private val pulseMetadata: PulseMetadata) {
    private val tonClient: TonClient = TonClient(
        cfg = TonClientConfig(networkId = networkId),
        projectId = projectId,
        pulseMetadata = pulseMetadata
    )

    // ==== INSTANCE METHODS ====

    fun generateKeypair(result: MethodChannel.Result) {
        val keyPair = tonClient.generateKeypair()
        result.success(mapOf("pk" to keyPair.pk, "sk" to keyPair.sk))
    }

    @SuppressLint("LongLogTag")
    fun generateKeypairFromTonMnemonic(params: Any?, result: MethodChannel.Result) {
        val dict = params as? Map<*, *> ?: return result.error("Ton.generateKeypairFromTonMnemonic", "Invalid parameters", null)
        Log.d("🤖 Ton.generateKeypairFromTonMnemonic", "params: $dict")

        val mnemonic = dict["mnemonic"] as? String ?: return result.error("Ton.generateKeypairFromTonMnemonic", "Missing mnemonic", null)
        val keyPair = tonClient.generateKeypairFromTonMnemonic(mnemonic = mnemonic)
        result.success(mapOf("pk" to keyPair.pk, "sk" to keyPair.sk))
    }

    @SuppressLint("LongLogTag")
    fun generateKeypairFromBip39Mnemonic(params: Any?, result: MethodChannel.Result) {
        val dict = params as? Map<*, *> ?: return result.error("Ton.generateKeypairFromBip39Mnemonic", "Invalid parameters", null)
        Log.d("🤖 Ton.generateKeypairFromBip39Mnemonic", "params: $dict")

        val mnemonic = dict["mnemonic"] as? String ?: return result.error("Ton.generateKeypairFromBip39Mnemonic", "Missing mnemonic", null)
        val keyPair = tonClient.generateKeypairFromBip39Mnemonic(mnemonic = mnemonic)
        result.success(mapOf("pk" to keyPair.pk, "sk" to keyPair.sk))
    }

    @SuppressLint("LongLogTag")
    fun getAddressFromKeypair(params: Any?, result: MethodChannel.Result) {
        val dict = params as? Map<*, *> ?: return result.error("getAddressFromKeypair", "Invalid parameters", null)
        Log.d("🤖 Ton.getAddressFromKeypair", "params: $dict")

        val sk = dict["sk"] as? String ?: return result.error("getAddressFromKeypair", "Missing sk", null)
        val pk = dict["pk"] as? String ?: return result.error("getAddressFromKeypair", "Missing pk", null)

        try {
            val keyPair = Keypair(sk, pk)
            val identity = tonClient.getAddressFromKeypair(keypair = keyPair)
            result.success(
                mapOf(
                    "friendlyEq" to identity.friendly,
                    "rawHex" to identity.rawHex,
                    "workchain" to identity.workchain.toHexString()
                )
            )
        } catch (e: Exception) {
            result.error("Ton.getAddressFromKeypair", e.message, null)
        }
    }

    fun signData(params: Any?, result: MethodChannel.Result) {
        val dict = params as? Map<*, *> ?: return result.error("signData", "Invalid parameters", null)
        Log.d("🤖 Ton.signData", "params: $dict")

        val text = dict["text"] as? String ?: return result.error("signData", "Missing text", null)
        val sk = dict["sk"] as? String ?: return result.error("signData", "Missing sk", null)
        val pk = dict["pk"] as? String ?: return result.error("signData", "Missing pk", null)

        try {
            val keyPair = Keypair(sk, pk)
            val response = tonClient.signData(text = text, keypair = keyPair)
            result.success(response)
        } catch (e: Exception) {
            result.error("Ton.signData", e.message, null)
        }
    }

    fun sendMessage(params: Any?, result: MethodChannel.Result) {
        val dict = params as? Map<*, *> ?: return result.error("sendMessage", "Invalid parameters", null)
        Log.d("🤖 Ton.sendMessage", "params: $dict")

        val networkId = dict["networkId"] as? String ?: return result.error("sendMessage", "Missing network", null)
        val from = dict["from"] as? String ?: return result.error("sendMessage", "Missing from", null)
        val validUntil = dict["validUntil"] as? Int ?: return result.error("sendMessage", "Missing validUntil", null)
        val messages = dict["messages"] as? List<*> ?: return result.error("sendMessage", "Missing messages", null)
        val sk = dict["sk"] as? String ?: return result.error("sendMessage", "Missing sk", null)
        val pk = dict["pk"] as? String ?: return result.error("sendMessage", "Missing pk", null)

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val keyPair = Keypair(sk, pk)
                val response = tonClient.sendMessage(
                    network = networkId,
                    from = from,
                    keypair = keyPair,
                    validUntil = validUntil.toUInt(),
                    messages = messages.toSendTxMessageList()
                )
                result.success(response)
            } catch (e: Exception) {
                Log.e("🤖 Ton.sendMessage", "Error: ${e.message}")
                result.error("Ton.sendMessage", e.message, null)
            }
        }
    }

    fun broadcastMessage(params: Any?, result: MethodChannel.Result) {
        val dict = params as? Map<*, *> ?: return result.error("broadcastMessage", "Invalid parameters", null)
        Log.d("🤖 Ton.broadcastMessage", "params: $dict")

        val from = dict["from"] as? String ?: return result.error("broadcastMessage", "Missing from", null)
        val validUntil = dict["validUntil"] as? UInt ?: return result.error("broadcastMessage", "Missing validUntil", null)
        val messages = dict["messages"] as? List<*> ?: return result.error("broadcastMessage", "Missing messages", null)
        val sk = dict["sk"] as? String ?: return result.error("broadcastMessage", "Missing sk", null)
        val pk = dict["pk"] as? String ?: return result.error("broadcastMessage", "Missing pk", null)

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val keyPair = Keypair(sk, pk)
                val response = tonClient.broadcastMessage(
                    from = from,
                    keypair = keyPair,
                    validUntil = validUntil,
                    messages = messages.toSendTxMessageList()
                )
                result.success(response)
            } catch (e: Exception) {
                Log.e("🤖 Ton.broadcastMessage", "Error: ${e.message}")
                result.error("Ton.broadcastMessage", e.message, null)
            }
        }
    }

    fun dispose() {
        // Clean up resources if TonClient exposes something like close()
        Log.d("🤖 Ton.dispose", "Disposing TonClient for network: $networkId")
    }

    // endregion

    // region ==== COMPANION OBJECT ====

    companion object {
        private val clients = mutableMapOf<String, Ton>()

        fun init(applicationContext: Context, params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_init", "Invalid parameters", null)
            Log.d("🤖 Ton.init", "params: $dict")

            val projectId = dict["projectId"] as? String ?: return errorMissing("projectId", params, result)
            val networkId = dict["networkId"] as? String ?: return errorMissing("networkId", params, result)
            val pulseMetadataDict = dict["pulseMetadata"] as? Map<*, *> ?: return errorMissing("pulseMetadata", params, result)

            val url = pulseMetadataDict["url"] as? String
            val packageName = applicationContext.packageName ?: pulseMetadataDict["packageName"] as? String ?: return errorMissing("packageName", params, result)
            val sdkVersion = pulseMetadataDict["sdkVersion"] as? String ?: return errorMissing("sdkVersion", params, result)
            val sdkPlatform = pulseMetadataDict["sdkPlatform"] as? String ?: return errorMissing("sdkPlatform", params, result)

            val pulseMetadata = PulseMetadata(url, packageName, sdkVersion, sdkPlatform)

            val ton = Ton(projectId, networkId, pulseMetadata)
            clients[networkId] = ton
            result.success(true)
        }

        private fun getClient(networkId: String, result: MethodChannel.Result): Ton? {
            val client = clients[networkId]
            if (client == null) result.error("TonClient", "networkId '$networkId' not found", null)
            return client
        }

        fun generateKeypair(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_generateKeypair", "Invalid parameters", null)
            Log.d("🤖 Ton.generateKeypair", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_generateKeypair", "Missing networkId", null)
            getClient(networkId, result)?.generateKeypair(result)
        }

        @SuppressLint("LongLogTag")
        fun generateKeypairFromTonMnemonic(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_generateKeypairFromTonMnemonic", "Invalid parameters", null)
            Log.d("🤖 Ton.generateKeypairFromTonMnemonic", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_generateKeypairFromTonMnemonic", "Missing networkId", null)
            getClient(networkId, result)?.generateKeypairFromTonMnemonic(dict, result)
        }

        @SuppressLint("LongLogTag")
        fun generateKeypairFromBip39Mnemonic(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_generateKeypairFromBip39Mnemonic", "Invalid parameters", null)
            Log.d("🤖 Ton.generateKeypairFromBip39Mnemonic", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_generateKeypairFromBip39Mnemonic", "Missing networkId", null)
            getClient(networkId, result)?.generateKeypairFromBip39Mnemonic(dict, result)
        }

        @SuppressLint("LongLogTag")
        fun getAddressFromKeypair(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_getAddressFromKeypair", "Invalid parameters", null)
            Log.d("🤖 Ton.getAddressFromKeypair", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_getAddressFromKeypair", "Missing networkId", null)
            getClient(networkId, result)?.getAddressFromKeypair(dict, result)
        }

        fun signData(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_signData", "Invalid parameters", null)
            Log.d("🤖 Ton.signData", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_signData", "Missing networkId", null)
            getClient(networkId, result)?.signData(dict, result)
        }

        fun sendMessage(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_sendMessage", "Invalid parameters", null)
            Log.d("🤖 Ton.sendMessage", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_sendMessage", "Missing networkId", null)
            getClient(networkId, result)?.sendMessage(dict, result)
        }

        fun broadcastMessage(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_broadcastMessage", "Invalid parameters", null)
            Log.d("🤖 Ton.broadcastMessage", "params: $dict")

            val networkId = dict["networkId"] as? String ?: return result.error("ton_broadcastMessage", "Missing networkId", null)
            getClient(networkId, result)?.broadcastMessage(dict, result)
        }

        fun dispose(params: Any?, result: MethodChannel.Result) {
            val dict = params as? Map<*, *> ?: return result.error("ton_dispose", "Invalid parameters", null)
            val networkId = dict["networkId"] as? String ?: return result.error("ton_dispose", "Missing networkId", null)
            clients.remove(networkId)?.dispose()
            result.success(true)
        }
    }
    // endregion
}

File: packages/reown_yttrium_utils/ios/Classes/Ton.swift
//
//  Ton.swift
//  reown_yttrium_utils
//
//  Created by Alfredo Rinaudo on 15/10/2025.
//
import Flutter
import UIKit
import YttriumUtilsWrapper

class Ton {
    private let tonClient: TonClient
    private let projectId: String
    private let networkId: String
    private let pulseMetadata: PulseMetadata
    
    private init(projectId: String, networkId: String, pulseMetadata: PulseMetadata) {
        self.projectId = projectId
        self.networkId = networkId
        self.pulseMetadata = pulseMetadata
        
        let config = TonClientConfig(networkId: networkId)
        self.tonClient = TonClient(cfg: config, projectId: projectId, pulseMetadata: pulseMetadata)
    }
    
    // MARK: - Instance methods
    
    private func generateKeypair(result: FlutterResult) {
        let keyPair = tonClient.generateKeypair()
        result([
            "pk": keyPair.pk,
            "sk": keyPair.sk
        ])
    }
    
    private func generateKeypairFromTonMnemonic(params: [String: Any], result: FlutterResult) {
        guard let mnemonic = params["mnemonic"] as? String else {
            return result(FlutterError(code: "Ton.generateKeypairFromTonMnemonic", message: "Missing mnemonic", details: nil))
        }
        
        do {
            let keyPair = try tonClient.generateKeypairFromTonMnemonic(mnemonic: mnemonic)
            result(["pk": keyPair.pk, "sk": keyPair.sk])
        } catch {
            result(FlutterError(code: "Ton.generateKeypairFromTonMnemonic", message: error.localizedDescription, details: nil))
        }
        
    }
    
    private func generateKeypairFromBip39Mnemonic(params: [String: Any], result: FlutterResult) {
        guard let mnemonic = params["mnemonic"] as? String else {
            return result(FlutterError(code: "Ton.generateKeypairFromBip39Mnemonic", message: "Missing mnemonic", details: nil))
        }
        
        do {
            let keyPair = try tonClient.generateKeypairFromBip39Mnemonic(mnemonic: mnemonic)
            result(["pk": keyPair.pk, "sk": keyPair.sk])
        } catch {
            result(FlutterError(code: "Ton.generateKeypairFromBip39Mnemonic", message: error.localizedDescription, details: nil))
        }
    }
    
    private func getAddressFromKeypair(params: [String: Any], result: FlutterResult) {
        guard
            let sk = params["sk"] as? String,
            let pk = params["pk"] as? String
        else {
            return result(FlutterError(code: "Ton.getAddressFromKeypair", message: "Missing sk or pk", details: nil))
        }
        
        do {
            let keyPair = Keypair(sk: sk, pk: pk)
            let identity = try tonClient.getAddressFromKeypair(keypair: keyPair)
            result([
                "friendlyEq": identity.friendly,
                "rawHex": identity.rawHex,
                "workchain": String(identity.workchain)
            ])
        } catch {
            result(FlutterError(code: "Ton.getAddressFromKeypair", message: error.localizedDescription, details: nil))
        }
    }
    
    private func signData(params: [String: Any], result: FlutterResult) {
        guard
            let text = params["text"] as? String,
            let sk = params["sk"] as? String,
            let pk = params["pk"] as? String
        else {
            return result(FlutterError(code: "Ton.signData", message: "Missing text/sk/pk", details: nil))
        }
        do {
            let keyPair = Keypair(sk: sk, pk: pk)
            let response = try tonClient.signData(text: text, keypair: keyPair)
            result(response)
        } catch {
            result(FlutterError(code: "Ton.signData", message: error.localizedDescription, details: nil))
        }
    }
    
    private func sendMessage(params: [String: Any], result: @escaping FlutterResult) {
        guard
            let networkId = params["networkId"] as? String,
            let from = params["from"] as? String,
            let validUntil = params["validUntil"] as? Int,
            let messages = params["messages"] as? [[String: Any]],
            let sk = params["sk"] as? String,
            let pk = params["pk"] as? String
        else {
            return result(FlutterError(code: "Ton.sendMessage", message: "Missing required fields", details: nil))
        }
        
        Task {
            do {
                let keyPair = Keypair(sk: sk, pk: pk)
                let txMessages = messages.toSendTxMessageList()
                let response = try await self.tonClient.sendMessage(
                    network: networkId,
                    from: from,
                    keypair: keyPair,
                    validUntil: UInt32(validUntil),
                    messages: txMessages
                )
                result(response)
            } catch {
                result(FlutterError(code: "Ton.sendMessage", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func broadcastMessage(params: [String: Any], result: @escaping FlutterResult) {
        guard
            let from = params["from"] as? String,
            let validUntil = params["validUntil"] as? UInt,
            let messages = params["messages"] as? [[String: Any]],
            let sk = params["sk"] as? String,
            let pk = params["pk"] as? String
        else {
            return result(FlutterError(code: "Ton.broadcastMessage", message: "Missing required fields", details: nil))
        }
        
        Task {
            do {
                let keyPair = Keypair(sk: sk, pk: pk)
                let txMessages = messages.toSendTxMessageList()
                let response = try await self.tonClient.broadcastMessage(
                    from: from,
                    keypair: keyPair,
                    validUntil: UInt32(validUntil),
                    messages: txMessages
                )
                result(response)
            } catch {
                result(FlutterError(code: "Ton.broadcastMessage", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    private func dispose() {
        print("🧹 Disposing TonClient for network \(networkId)")
    }
    
    // MARK: - Static instance manager
    
    private static var clients: [String: Ton] = [:]
    
    static func initialize(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any] else {
            result(FlutterError(code: "Ton.initialize", message: "Invalid parameters: not a map", details: params))
            return
        }
        
        guard
            let networkId = dict["networkId"] as? String,
            let projectId = dict["projectId"] as? String,
            let pulseMetadataDict = dict["pulseMetadata"] as? [String: Any],
            let sdkVersion = pulseMetadataDict["sdkVersion"] as? String,
            let sdkPlatform = pulseMetadataDict["sdkPlatform"] as? String
        else {
            result(FlutterError(code: "Ton.initialize", message: "Missing required fields", details: dict))
            return
        }
        
        let url = pulseMetadataDict["url"] as? String
        let bundleId = pulseMetadataDict["bundleId"] as? String
        
        let pulseMetadata = PulseMetadata(url: url,
                                          bundleId: bundleId ?? Bundle.main.bundleIdentifier,
                                          sdkVersion: sdkVersion,
                                          sdkPlatform: sdkPlatform)
        
        let ton = Ton(projectId: projectId, networkId: networkId, pulseMetadata: pulseMetadata)
        clients[networkId] = ton
        result(true)
    }
    
    private static func getClient(for networkId: String, result: FlutterResult) -> Ton? {
        guard let client = clients[networkId] else {
            result(FlutterError(code: "TonClient", message: "networkId '\(networkId)' not found", details: nil))
            return nil
        }
        return client
    }
    
    // MARK: - Static entrypoints
    
    public static func generateKeypair(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.generateKeypair", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.generateKeypair", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.generateKeypair(result: result)
    }
    
    public static func generateKeypairFromTonMnemonic(_ params: Any, result: FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.generateKeypairFromTonMnemonic", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.generateKeypairFromTonMnemonic", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.generateKeypairFromTonMnemonic(params: dict, result: result)
    }
    
    public static func generateKeypairFromBip39Mnemonic(_ params: Any, result: FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.generateKeypairFromBip39Mnemonic", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.generateKeypairFromBip39Mnemonic", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.generateKeypairFromBip39Mnemonic(params: dict, result: result)
    }
    
    public static func getAddressFromKeypair(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.getAddressFromKeypair", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.getAddressFromKeypair", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.getAddressFromKeypair(params: dict, result: result)
    }
    
    public static func signData(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.signData", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.signData", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.signData(params: dict, result: result)
    }
    
    public static func sendMessage(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.sendMessage", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.sendMessage", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.sendMessage(params: dict, result: result)
    }
    
    public static func broadcastMessage(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.broadcastMessage", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.broadcastMessage", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.broadcastMessage(params: dict, result: result)
    }
    
    public static func dispose(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let networkId = dict["networkId"] as? String else {
            result(FlutterError(code: "Ton.dispose", message: "Invalid parameters", details: params))
            return
        }
        
        guard let client = getClient(for: networkId, result: result) else {
            result(FlutterError(code: "Ton.dispose", message: "TonClient for networkId \(networkId) not found", details: nil))
            return
        }
        
        client.dispose()
        clients.removeValue(forKey: networkId)
        
        result(true)
    }
}
